### PR TITLE
Get closest free buildings

### PR DESCRIPTION
**Pull request by JordanRoth**
This PR also includes all of the changes from this (now-defunct) [PR](https://github.com/esmason/roomr/pull/2). Please see that PR for notes on many changes.

In addition, the ability to call a single function to get the n closest buildings (where each building has at least one free room) to some given point at some given time on some given day has been added. No front-end work was done.

### CODE REVIEW

File: server/database_access/retrieve-buildings.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * @param time
+ * @param day
+ * @returns {Array.<T>}
+ */
+export function getClosestAvailableBuildings(numBuildings, lat, lon, time, day) {
+    const availableBuildingsArray = [];
+    Buildings.find({}).forEach(function (building) {
+        let availableRooms = [];
+        building.Rooms().forEach(function(room) {
+            let clashingRoomSlot = false;
```
Comment (IMCampbell) [inline server/database_access/retrieve-buildings.js]: Rename clash to something that implies the room is occupied, not clashing with another roomslot.

File: server/database_access/retrieve-buildings.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    return (startTime <= time && time < endTime && roomSlot.days.indexOf(day) != -1);
+}
+
+/**
+ * Takes a building and its available rooms and creates a new object to send to the front end.
+ * @param building
+ * @param availableRooms
+ * @returns {{_id: *, name: *, latitude: *, longitude: *, availableRooms: *}}
+ */
+function getReturnBuildingObject(building, availableRooms) {
```
Comment (IMCampbell) [inline server/database_access/retrieve-buildings.js]: Maybe rename to getFrontEndBuildingObject or getDisplayBuildingObject?

File: server/main.js [hunk 3]
```diff
@@ -1 +1,25 @@
-import '../imports/api/times.js';
+import '../imports/api/times.js'
```
Comment (IMCampbell) [inline server/main.js]: ;

File: server/main.js [hunk 4]
```diff
@@ -1 +1,25 @@
...
+import '../imports/database/collections';
+import './database_access/retrieve-buildings.js'
```
Comment (IMCampbell) [inline server/main.js]: ;

File: server/utilities/distance-utils.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+export function getDistanceFromLatLonInKm(lat1,lon1,lat2,lon2) {
```
Comment (IMCampbell) [inline server/utilities/distance-utils.js:5]: Comment code, and credit http://stackoverflow.com/questions/27928/calculate-distance-between-two-latitude-longitude-points-haversine-formula

**Review by IMCampbell — CHANGES_REQUESTED** [review]

File: server/main.js [hunk 6]
```diff
@@ -1 +1,25 @@
-import '../imports/api/times.js';
+import '../imports/api/times.js'
+import '../imports/database/collections';
```
Comment (esmason) [inline server/main.js]: add helper function to collections.js instead of using import to run file

File: server/main.js (hunk 4 above)
Comment (esmason) [inline server/main.js]: try removing first 3 lines of file

File: server/database_initialization/database-construction.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      "Wed",
+      "Fri"
+    ],
+    "startTime": "12:00",
+    "endTime": "13:00",
+    "building": "Buchanan",
+    "roomID": "B313",
+    "sectionID": "ANTH 201A 001",
+    "address": "1866 Main Mall"
+   {
```
Comment (esmason) [inline server/database_initialization/database-construction.js]: { -> }

File: server/database_initialization/database-construction.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ */
+export function persistScraperRoomSlots(roomSlots) {
+    const newBuildings = [];
```
Comment (esmason) [inline server/database_initialization/database-construction.js]: call array "Buildings"

File: server/database_initialization/database-construction.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    console.log("Database Initialized!");
+}
+
+/**
+ * Takes a roomSlot and gets the relevant building from the buildings collection, or adds it if it didn't exist.
+ * @param roomSlot
+ * @param newBuildings: an array of the buildings that did not already belong to the buildings collection.
+ * @returns building
+ */
+function parseBuilding(roomSlot, newBuildings) {
```
Comment (esmason) [inline server/database_initialization/database-construction.js]: --> `function getOrCreateBuilding(roomSlot, Buildings)`

File: server/database_initialization/database-construction.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    return building;
+}
+
+/**
+ * Takes a roomSlot and gets the relevant room from the rooms collection, or adds it if it didn't exist.
+ * @param roomSlot
+ * @param building
+ * @returns room
+ */
+function parseRoom(roomSlot, building) {
```
Comment (esmason) [inline server/database_initialization/database-construction.js]: rename `function getOrCreateRoom(roomSlot, building)`

File: server/database_initialization/database-construction.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    return room;
+}
+
+/**
+ * Takes a roomSlot and parses it into a roomr roomSlot object and stores it into the roomSlot collection.
+ * @param roomSlot
+ * @param building
+ * @param room
+ */
+function parseRoomSlot(roomSlot, building, room) {
```
Comment (esmason) [inline server/database_initialization/database-construction.js]: rename `createRoomSlot`

File: server/utilities/date-time-utils.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * In order to parse a string as a time object, it needs to be in the format of hh:mm. Many times come in the format of
+ * h:mm (e.g 9:00) so we need to pad the time (e.g. 09:00).
+ * @param time
+ * @returns (LocalTime)
+ */
+export function stringToLocalTime(time) {
+    const split = time.split(":");
+    if (split[0].length == 1) {
+        return LocalTime.parse((`0${ split[0] }:${ split[1] }`))
```
Comment (esmason) [inline server/utilities/date-time-utils.js]: change `('0${ split[0] }:${ split[1] }')` to `'0'.concat(time)`

File: server/database_initialization/geocoder.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    return updateObjects;
+}
+
+/**
+ * Takes a singular building and the correct latitude and longitude and updates the building.
+ * @param newLat
+ * @param newLon
+ * @param id
+ */
+function updateBuildingWithLatLon(newLat, newLon, id) {
```
Comment (esmason) [inline server/database_initialization/geocoder.js]: validate lat lon (potential add helper fn to Building)

**Comment by esmason** [comment]
recommended database changes

- Separate each collection into it's own file
- Move all insert and update functionality to custom helper functions
- schema can be validated in helper functions

**Review by esmason — COMMENTED** [review]
recommended database changes

  -  Separate each collection into it's own file
 -   Move all insert and update functionality to custom helper functions
 -  schema can be validated in helper functions

File: server/database_initialization/database-construction.js [hunk 14]
```diff
@@ -23,26 +25,23 @@ import { RoomSlots, Rooms, Buildings } from '/imports/database/collections';
...
  * Takes a roomSlot and gets the relevant building from the buildings collection, or adds it if it didn't exist.
- * @param roomSlot
- * @param newBuildings: an array of the buildings that did not already belong to the buildings collection.
- * @returns building
  */
-function parseBuilding(roomSlot, newBuildings) {
+function getOrCreateBuilding(roomSlot, buildings) {
     let building = Buildings.findOne({name: roomSlot['building']});
     if (typeof building === 'undefined') {
         let newBuilding = {
```
Comment (IMCampbell) [inline server/database_initialization/database-construction.js]: should be buildings?

File: server/database_initialization/database-construction.js [hunk 15]
```diff
@@ -23,26 +25,23 @@ import { RoomSlots, Rooms, Buildings } from '/imports/database/collections';
...
 
 /**
  * Takes a roomSlot and gets the relevant building from the buildings collection, or adds it if it didn't exist.
- * @param roomSlot
- * @param newBuildings: an array of the buildings that did not already belong to the buildings collection.
- * @returns building
  */
-function parseBuilding(roomSlot, newBuildings) {
+function getOrCreateBuilding(roomSlot, buildings) {
     let building = Buildings.findOne({name: roomSlot['building']});
```
Comment (IMCampbell) [inline server/database_initialization/database-construction.js]: This could be refactored to be clearer, too many building-like variables.

**Review by IMCampbell — CHANGES_REQUESTED** [review]

**Review by IMCampbell — APPROVED** [review]
Neato!

**Pull request merged**